### PR TITLE
Grant TLS tag privilege to gRPC client pseudo-node

### PR DESCRIPTION
**Pull request by tiziano88**
Replace the certificate SAN name field in the label, with a field that
represents the authority (host:port) of the remote endpoint instead.

Make `run_event_loop` more type safe by passing in a `Reciever<T>` where
`T` is the same type of the `Node<T>` that actually handles the
messages.

Create an explicit `GrpcServerInit` message to pass to a gRPC server
pseudo-node, instead of manually passing handles to it in an
unstructured way.

Remove unused `address` field from `GrpcClientConfiguration` protobuf
message.

Split aggregator example in two nodes, so that the gRPC worker node may
be assigned a different label in the future, to reflect the fact that it
is handling non-public data.

Add explicit scheme to gRPC client URL in aggregator example (not sure
how / whether it worked previously, Tonic complains).

Make aggregator client submit a meaningful number of samples in order to
reach the aggregation threshold, instead of just one.

Ref #814

# Checklist

- [ ] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [ ] I have written tests that cover the code changes.
  - [ ] I have checked that these tests are run by
        [Cloudbuild](/cloudbuild.yaml)
  - [ ] I have updated [documentation](/docs/) accordingly.
  - [ ] I have raised an [issue](https://github.com/project-oak/oak/issues) to
        cover any TODOs and/or unfinished work.
- [ ] Pull request includes prototype/experimental work that is under
      construction.

### CODE REVIEW

**Comment by tiziano88** [comment]
FYI tests seem to be failing because of a Wasm node creation:

```
[2020-07-01T20:53:34Z WARN  oak_runtime::runtime] could not create node: IncorrectWebAssemblyModuleName
[2020-07-01T20:53:34Z ERROR oak_runtime::node::wasm] test: node_create(): Could not create node: ErrInvalidArgs
[2020-07-01T20:53:34Z ERROR oak_runtime::node::logger] panic occurred in file 'aggregator/module/rust/src/lib.rs' at line 173: could not create gRPC worker node: ErrInvalidArgs
```

Originally started from here:

https://github.com/project-oak/oak/blob/2cf69a76494d21189c171263679a9a4199866e29/examples/aggregator/module/rust/tests/integration_test.rs#L46-L47

it seems that in integration tests, the module is called `aggregator`, but when running as an example it is called `app`?

https://github.com/project-oak/oak/blob/2cf69a76494d21189c171263679a9a4199866e29/examples/aggregator/module/rust/tests/integration_test.rs#L27

I don't fully understand the flow though, @daviddrysdale could you confirm this is the case? Where does the name come from in the integration tests?

**Comment by daviddrysdale** [comment]
The mismatch between `wasm_module_name = app` in `config.textproto` and `MODULE_CONFIG_NAME = aggregator` in `integration_test.rs` does seem odd/wrong, but it turns out that the same mismatch is present in almost every example.   

However, I guess this is one of the few examples that creates a second Node at runtime, so maybe that does an additional check somehow?  (The `abitest` example also creates extra Nodes at runtime, but I think it has matching module names.)

File: oak/proto/label.proto [hunk 1]
```diff
@@ -66,16 +66,7 @@ message WebAssemblyModuleTag {
...
-  // In general a certificate may have multiple SANs; an HTTP or gRPC client pseudo-node connected
-  // to a remote endpoint with multiple SANs is able to declassify data with a TlsEndpointTag
-  // referring to any of these SANs.
-  //
-  // The certificate is validated by the Oak Runtime using the set of Certificate Authorities (CA)
-  // that are available to it.
-  //
-  // See https://tools.ietf.org/html/rfc5280#section-4.2.1.6
-  string certificate_subject_alternative_name = 1;
+  // The TLS authority (host:port) of the remote endpoint, which is validated by the Oak Runtime
```
Comment (conradgrobler) [inline oak/proto/label.proto:69]: nit: I think host is sufficient for trust. Port is probably not needed.

**Comment by tiziano88** [comment]
> The mismatch between `wasm_module_name = app` in `config.textproto` and `MODULE_CONFIG_NAME = aggregator` in `integration_test.rs` does seem odd/wrong, but it turns out that the same mismatch is present in almost every example.   
> 
> However, I guess this is one of the few examples that creates a second Node at runtime, so maybe that does an additional check somehow?  (The `abitest` example also creates extra Nodes at runtime, but I think it has matching module names.)

Well yes, because creating a new node requires the "fully qualified" module + function name, so in this case there is no module name that works for both. I tried changing the const to "app" but then it does not find the module in the first place (I think), it must be derived from the package name or something else. Will look more closely.

Also I'm thinking we should have a way of creating a node with a fn from the same module as the caller, perhaps just specifying an empty string for the module name, or a special reserved value like "self".

File: examples/aggregator/client/aggregator.cc [hunk 2]
```diff
@@ -103,8 +103,10 @@ int main(int argc, char** argv) {
     values.push_back(value);
   }
 
-  // Submit data sample.
-  submit_sample(stub.get(), bucket, indices, values);
+  // Submit data sample as many times as it takes to reach the hardcoded threshold.
+  for (int i = 0; i < 5; i++) {
+    submit_sample(stub.get(), bucket, indices, values);
+  }
```
Comment (tiziano88) [inline examples/aggregator/client/aggregator.cc:109]: @ipetr0v let me know what you think of this. I can remove it if you prefer, but there must be a mechanism somewhere to allow for testing the actual logic that we are interested in (even if it's just the `run_example` script calling this binary 5 times).

File: examples/aggregator/module/rust/src/lib.rs [hunk 3]
```diff
@@ -106,7 +106,7 @@ impl AggregatorNode {
         //     integrity_tags: vec![],
         // };
         match oak::grpc::client::Client::new_with_label(
-            &oak::node_config::grpc_client("127.0.0.1:8888"),
+            &oak::node_config::grpc_client("https://localhost:8888"),
```
Comment (tiziano88) [inline examples/aggregator/module/rust/src/lib.rs:109]: @ipetr0v this is the part that I don't understand how it could work before (did it?).

File: oak/proto/label.proto (hunk 1 above)
Comment (tiziano88) [reply oak/proto/label.proto:69]: Yes I think you are right, but since we have the port, it may help prevent some confused deputy problem in which someone manages to get a web page on some third party server that also has some gRPC server running, and get data out. Not entirely sure, we can easily take it out in the future if we think it's safe anyways.

File: examples/aggregator/module/rust/src/lib.rs (hunk 3 above)
Comment (ipetr0v) [reply examples/aggregator/module/rust/src/lib.rs:109]: I think it didn't. Strange that CI didn't pick it up.

File: examples/aggregator/client/aggregator.cc (hunk 2 above)
Comment (ipetr0v) [reply examples/aggregator/client/aggregator.cc:109]: I think it's better than it was before.

**Comment by tiziano88** [comment]
Please also review / approve when you have some time @ipetr0v .

File: examples/aggregator/module/rust/src/lib.rs [hunk 4]
```diff
@@ -157,9 +157,26 @@ impl Aggregator for AggregatorNode {
...
+});
+
+oak::entrypoint!(oak_main => |_in_channel| {
+    oak::logger::init_default();
+    let grpc_channel =
+        oak::grpc::server::init("[::]:8080").expect("could not create gRPC server pseudo-Node");
+    let (invocation_sender, invocation_receiver) =
+        oak::io::channel_create::<oak::grpc::Invocation>()
+            .expect("could not create gRPC invocation channel");
+    oak::node_create(
```
Comment (ipetr0v) [inline examples/aggregator/module/rust/src/lib.rs:173]: Does it mean that the initial Node and the worker Node are defined in the same Wasm file?
Is it supposed to be like this, or it's better to have Node=module?

File: oak/server/rust/oak_runtime/src/node/grpc/client.rs [hunk 5]
```diff
@@ -57,6 +59,22 @@ fn check_uri(uri: &Uri) -> Result<(), ConfigurationError> {
         .ok_or(ConfigurationError::NoHostElement)
 }
 
+fn grpc_client_node_privilege(uri: &Uri) -> NodePrivilege {
+    if uri.scheme() == Some(&http::uri::Scheme::HTTPS) {
+        // Authority is the host:port portion of the endpoint name.
+        if let Some(authority) = uri.authority() {
+            NodePrivilege::new(
+                hashset! { oak_abi::label::tls_endpoint_tag(&authority.as_str()) },
+                hashset! { oak_abi::label::tls_endpoint_tag(&authority.as_str()) },
```
Comment (ipetr0v) [inline oak/server/rust/oak_runtime/src/node/grpc/client.rs:68]: Why is this string copied two times?

File: oak/common/label.cc [hunk 6]
```diff
@@ -59,6 +59,13 @@ oak::label::Label WebAssemblyModuleLabel(const std::string& web_assembly_module_
   return label;
 }
 
+oak::label::Label TlsEndpointLabel(const std::string& authority) {
```
Comment (ipetr0v) [inline oak/common/label.cc:62]: Does authority allow wildcards `*`?

File: sdk/rust/oak/src/grpc/server.rs [hunk 7]
```diff
@@ -16,27 +16,68 @@
 
 //! Functionality to help Oak Nodes create gRPC server pseudo-Nodes.
 
-use crate::{OakStatus, ReadHandle};
+use crate::{grpc::Invocation, io::Receiver, OakStatus};
+
+struct GrpcServerInit {
```
Comment (ipetr0v) [inline sdk/rust/oak/src/grpc/server.rs:21]: Why is this message needed?

File: examples/aggregator/module/rust/src/lib.rs (hunk 4 above)
Comment (tiziano88) [reply examples/aggregator/module/rust/src/lib.rs:173]: In this case there is no use case in splitting them, so it is much more convenient to do it in the same module. But of course, if we plan to have a reusable aggregator module, we should consider splitting them in the future.

File: oak/common/label.cc (hunk 6 above)
Comment (tiziano88) [reply oak/common/label.cc:62]: No, currently it's literal `host:port`.

File: oak/server/rust/oak_runtime/src/node/grpc/client.rs (hunk 5 above)
Comment (tiziano88) [reply oak/server/rust/oak_runtime/src/node/grpc/client.rs:68]: One is the privilege to declassify, and the other is the privilege to endorse. Basically confidentiality + integrity. In this case (and in many cases in practice) they happen to coincide.

File: sdk/rust/oak/src/grpc/server.rs (hunk 7 above)
Comment (tiziano88) [reply sdk/rust/oak/src/grpc/server.rs:21]: To avoid passing handles around manually. Currently the interaction is very low level, we pass handles around using channel_write, but that's not really how channels are meant to be used. That's why I added a TODO on the gRPC server pseudo-node impl, to migrate that to using a proper encapsulated type, rather than reading individual handles and having assertions around them that they are of the correct number and shape. Basically currently we are doing the equivalent of reading bytes and putting them together, instead of using a proper serialization format like protobuf.

TL;DR: type safety == good; manually manipulating bytes + handles == bad

File: oak/server/rust/oak_runtime/src/node/grpc/client.rs (hunk 5 above)
Comment (ipetr0v) [reply oak/server/rust/oak_runtime/src/node/grpc/client.rs:68]: Should we use different types so it will be harder to make mistakes when setting labels?
Or confidentiality and integrity are always the same?

File: oak/common/label.cc (hunk 6 above)
Comment (ipetr0v) [reply oak/common/label.cc:62]: Does it propagate to subdomains?

Comment (tiziano88) [reply oak/common/label.cc:62]: No, it is just `host:port` exactly as it appears in the URL.

File: oak/server/rust/oak_runtime/src/node/grpc/client.rs (hunk 5 above)
Comment (tiziano88) [reply oak/server/rust/oak_runtime/src/node/grpc/client.rs:68]: Maybe I should just remove the constructor, and build the struct directly with the field names, so at least the values are next to their description.

They are of the same type, so I don't think we can use different types though.

File: sdk/rust/oak/src/grpc/server.rs [hunk 8]
```diff
@@ -16,27 +16,68 @@
...
+    let (invocation_sender, invocation_receiver) =
+        crate::io::channel_create::<Invocation>().expect("Couldn't create gRPC invocation channel");
+
+    let grpc_server_init = GrpcServerInit { invocation_sender };
+    init_sender
+        .send(&grpc_server_init)
+        .expect("Could not send init message to gRPC server pseudo-node");
+    init_sender
+        .close()
+        .expect("Couldn't close init message channel to gRPC server pseudo-node");
```
Comment (daviddrysdale) [inline sdk/rust/oak/src/grpc/server.rs:80]: I think this induces a handle leak &ndash; the local Node's copy of the `invocation_sender` handle is not closed after it has been sent to the gRPC server p-N, and it is not saved anywhere so the handle value is lost.

Comment (daviddrysdale) [reply sdk/rust/oak/src/grpc/server.rs:80]: Raised as #1358.
